### DISCUSSION TITLE
为啥 Paxos 这么难？ - XX’s Blog

### DISCUSSION THREAD

**Discussion by giscus** [xxchan/xxchan.github.io#28] [Announcements]
# 为啥 Paxos 这么难？ - XX’s Blog

English Version

https://xxchan.github.io/cs/2022/02/09/paxos-hard-zh.html

**Comment by Xuanwo** [comment]
我确实花了很多时间才反应过来 Paxos 的 Paper 只讨论（证明）了 “分布式环境下如何就单个 value 达成一致“  这一件事（

**Reply by xxchan** [reply]
所以像理论人一样，讨论问题的时候先用数学语言定义好抽象/接口就很好（
（看理论教材/课程的时候就很爽

**Comment by im-rinnin** [comment]
> Paxos 则完全是 peer-to-peer
paxos-simple 论文里面说如果要保证progress，是需要leader的
> To guarantee progress, a distinguished proposer must be selected as the
only one to try issuing proposals.

**Reply by xxchan** [reply]
OK，是我疏忽了感谢指正。

但是其实是这样的：给 Paxos 加上 leader 非常容易，就只需要加个 if 判断，只有自己是 leader 的时候才能 propose 就行了。这也是论文中说的 “However, safety is ensured regardless of the success or failure of the election” 的意思。

相反，Raft 则是完全围绕 leader 设计的。

**Reply by drmingdrmer** [reply]
其实可以认为raft和paxos的leader是完全一样的.

vote 对等于 phase-1, append-log 对等于 phase-2.

**Reply by xxchan** [reply]
这么说倒也确实可以

（所以懂了 Paxos 以后再回头看 Raft 又变得更简单了

**Reply by drmingdrmer** [reply]
反正diss raft的我都顶...

**Comment by vickiegpt** [comment]
感觉这句话`request promise before write`有点循环定义，promise实现也可是基于共识算法的。

**Reply by ywgrit** [reply]
orz，这也可以看见佬
